gobot/handlers/pull_request_create.go

```diff
@@ -0,0 +1,483 @@
+		PullRequestCreateHandler: *prc,
+	}
+
+	prTask.branchName = prc.generateBranchName()
+
+	// Clone the taxonomy repo
+	repo, err := prTask.cloneTaxonomyRepo()
+	if err != nil {
+		prc.Logger.Errorf("Error cloning taxonomy repo %v ", err)
+		http.Error(w, "Error cloning taxonomy repo", http.StatusInternalServerError)
+		return
+	}
+
+	// Checkout branch for the pull request
+	wt, err := prTask.checkoutBranch(repo)
+	if err != nil {
+		prc.Logger.Errorf("Error checking out branch %v ", err)
+		http.Error(w, "Error checking out branch", http.StatusInternalServerError)
+		return
+	}
+
+	// Create commit for the pull request
+	commitSha, err := prTask.createKnowledgeCommit(wt, requestData)
+	if err != nil {
+		prc.Logger.Errorf("Error creating knowledge commit %v ", err)
+		http.Error(w, "Error creating knowledge commit", http.StatusInternalServerError)
+		return
+	}
+
+	// Push the commit to the taxonomy repo
+	err = prTask.pushCommit(repo)
+	if err != nil {
+		prc.Logger.Errorf("Error pushing knowledge commit %v ", err)
+		http.Error(w, "Error pushing knowledge commit", http.StatusInternalServerError)
+		return
+	}
+
+	// Create the pull request of the pushed branch to taxonomy main branch
+	pr, err := prTask.createPullRequest(KnowledgeStr, requestData.Task_description, requestData.Task_details)
+	if err != nil {
+		prc.Logger.Errorf("Error creating knowledge pull request %v ", err)
+		http.Error(w, "Error creating knowledge pull request", http.StatusInternalServerError)
+		return
+	}
+
+	prc.Logger.Infof("Pull request (%s) created successfully for knowledge with commit sha: %s", pr.GetHTMLURL(), commitSha)
+
+	w.WriteHeader(http.StatusOK)
+	fmt.Fprintf(w, "Pull request (%s) created successfully for the knowledge", pr.GetHTMLURL())
+
+	//Delete the directory after the pull request is created
+	err = os.RemoveAll(prTask.branchName)
+	if err != nil {
+		prc.Logger.Errorf("Error deleting branch directory %v ", err)
+		return
+	}
+}
+
+func (prt *PullRequestTask) cloneTaxonomyRepo() (*git.Repository, error) {
+	sugar := prt.Logger.With("user_name", prt.GithubUsername)
+
+	workDir, err := os.Getwd()
+	if err != nil {
+		return nil, fmt.Errorf("could not get working directory: %v", err)
+	}
+	taxonomyDir := path.Join(workDir, prt.branchName, TaxonomyPath)
+
+	// Check if the taxonomy directory exists and delete if it does
+	if _, err := os.Stat(taxonomyDir); !os.IsNotExist(err) {
+		sugar.Warn("Taxonomy directory already exists, deleting")
+		err = os.RemoveAll(taxonomyDir)
+		if err != nil {
+			return nil, fmt.Errorf("could not delete taxonomy directory: %v", err)
+		}
+	}
+
+	var r *git.Repository
+	if _, err := os.Stat(taxonomyDir); os.IsNotExist(err) {
+		sugar.Warnf("Taxonomy directory does not exist, cloning from %s", prt.TaxonomyRepo)
+		r, err = git.PlainClone(taxonomyDir, false, &git.CloneOptions{
+			URL: prt.TaxonomyRepo,
+			Auth: &githttp.BasicAuth{
+				Username: prt.GithubUsername,
+				Password: prt.GithubToken,
+			},
```

Comment:
Question, we push commits later on. Will this work over an https remote? I was under the impression even with a PAT no pushing over https is allowed...

Reply:
It does allow you to create a pull request/commit if write permissions for PR is enabled. Bot does use the https to push the changes.

Reply:
Got it, thank you for the clarification!